File: 0114-flatten-binary-tree-to-linked-list/0114-flatten-binary-tree-to-linked-list.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def flatten(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        if not root:
            return root
        left = self.flatten(root.left)
        right = self.flatten(root.right)
        root.left = None
        root.right = left
        cur = root
        while cur.right:
            cur = cur.right
        cur.right = right
        return root

File: 0114-flatten-binary-tree-to-linked-list/0114-flatten-binary-tree-to-linked-list.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

/**
 Do not return anything, modify root in-place instead.
 */
function flatten(root: TreeNode | null): void {
    go(root)
    
    function go(root: TreeNode | null): TreeNode | null {
        if (!root) return root
        const left = go(root.left)
        const right = go(root.right)
        let cur = root
        cur.left = null
        cur.right = left
        while (cur.right) cur = cur.right
        cur.right = right
        return root
    }
};